fix(ivy): ngcc - prefer JavaScript source files when resolving module…

JoostK opened the pull request:
… imports

Packages that do not follow APF may have the declaration files in the same
directory as one source format, typically ES5. This is problematic for ngcc,
as it needs to create a TypeScript program with all JavaScript sources of
an entry-point, whereas TypeScript's module resolution mechanism would have
resolved an internal module import to the external facing .d.ts declaration
file, instead of the JavaScript source file. This behavior results in the
program to be analysed being incomplete.

This commit introduces a custom compiler host that recognizes the above
scenario and rewires the resolution of a .d.ts declaration file to its
JavaScript counterpart, if applicable.

Fixes #29939

JoostK committed:
fix(ivy): ngcc - prefer JavaScript source files when resolving module imports

Packages that do not follow APF may have the declaration files in the same
directory as one source format, typically ES5. This is problematic for ngcc,
as it needs to create a TypeScript program with all JavaScript sources of
an entry-point, whereas TypeScript's module resolution mechanism would have
resolved an internal module import to the external facing .d.ts declaration
file, instead of the JavaScript source file. This behavior results in the
program to be analysed being incomplete.

This commit introduces a custom compiler host that recognizes the above
scenario and rewires the resolution of a .d.ts declaration file to its
JavaScript counterpart, if applicable.

Fixes #29939

petebacondarwin commented:
I think we should land https://github.com/angular/angular/pull/29643 and then consider using the `NgccCompilerHost` and the `ModuleResolver` to do this work.

JoostK commented on packages/compiler-cli/ngcc/src/packages/entry_point_bundle.ts:
```diff
@@ -47,17 +47,55 @@ export function makeEntryPointBundle(
+      host.getCurrentDirectory(), file => host.getCanonicalFileName(file));
+  host.resolveModuleNames = (moduleNames, containingFile, reusedNames, redirectedReference) => {
+    return moduleNames.map(moduleName => {
+      const {resolvedModule} = ts.resolveModuleName(
```
This is almost the same code as what TS uses by default, with one difference: TypeScript's default logic has an additional caching layer, besides the module resolution cache:

See `loadWithLocalCache`:
https://github.com/Microsoft/TypeScript/blob/b534fb4849eca0a792199fb6c0cb8849fece1cfd/src/compiler/program.ts#L409-L426

used here:
https://github.com/Microsoft/TypeScript/blob/b534fb4849eca0a792199fb6c0cb8849fece1cfd/src/compiler/program.ts#L653-L657

MathewBerg commented:
Until/if this PR lands, is there a suggested workaround that can be done locally? I'd love to test IVY but use angularfire2 extensively.

petebacondarwin commented:
Now that #29643 has landed we should be able to move this PR along.

JoostK committed:
fixup! fix(ivy): ngcc - prefer JavaScript source files when resolving module imports

JoostK commented:
I rebased this and derived a custom compiler host based on `NgccCompilerHost`. I did not use the `ModuleResolver` right now, as it introduces some things we would need to think about:

1) For `ResolvedExternalModule` we'd have to determine the typings location, i.e. load the `package.json` file and resolve its `typings` key. If it doesn't have it, we should instead use `main` (or others?) to resolve the `.js` source file instead?
2) For `ResolvedDeepImport` we also need to resolve the `typings` key in `package.json` and derive the correct `.d.ts` file use the deeply imported path and and the root of the typings directory.

I can't really foresee if the above would be at all accurate/cover our use-cases.

petebacondarwin commented on packages/compiler-cli/ngcc/src/packages/ngcc_compiler_host.ts:
```diff
@@ -64,3 +64,43 @@ export class NgccCompilerHost implements ts.CompilerHost {
+ * would otherwise let TypeScript prefer the .d.ts file instead of the JavaScript source file.
+ */
+export class NgccSourcesCompilerHost extends NgccCompilerHost {
+  readonly cache = ts.createModuleResolutionCache(
```
Why `readonly` rather than `private`?

JoostK replied:
I should have made it `private`. The `readonly` is a habit, I just forget the `private` ;)

petebacondarwin commented on packages/compiler-cli/ngcc/src/packages/ngcc_compiler_host.ts:
```diff
@@ -64,3 +64,43 @@ export class NgccCompilerHost implements ts.CompilerHost {
+      // If the module request originated from a relative import in a JavaScript source file,
+      // TypeScript may have resolved the module to its .d.ts declaration file if the .js source
+      // file was in the same directory. This is undesirable, as we need to have the actual
+      // JavaScript being present in the program. This logic recognizes this scenario and rewires
```
```suggestion
      // JavaScript being present in the program. This logic recognizes this scenario and rewrites
```

petebacondarwin commented on packages/compiler-cli/ngcc/src/packages/ngcc_compiler_host.ts:
```diff
@@ -64,3 +64,43 @@ export class NgccCompilerHost implements ts.CompilerHost {
+      // the resolved .d.ts declaration file to its .js counterpart, if it exists.
+      if (resolvedModule !== undefined && resolvedModule.extension === ts.Extension.Dts &&
+          containingFile.endsWith('js') &&
+          (moduleName.startsWith('./') || moduleName.startsWith('../')) &&
```
Use `isRelativePath()` from `utils.ts`

petebacondarwin commented on packages/compiler-cli/ngcc/src/packages/ngcc_compiler_host.ts:
```diff
@@ -64,3 +64,43 @@ export class NgccCompilerHost implements ts.CompilerHost {
+      if (resolvedModule !== undefined && resolvedModule.extension === ts.Extension.Dts &&
+          containingFile.endsWith('js') &&
+          (moduleName.startsWith('./') || moduleName.startsWith('../')) &&
+          resolvedModule.resolvedFileName.startsWith(this.entryPointPath)) {
```
I think that this is not correct: Source files can exist in a folder that is outside of the entry-point folder (but in the same package) while still being part of the source for that entry-point. Angular Material is an example of this.

Nor is it needed, I believe: if the import is "relative" then it would be breaking the node.js module model to relatively import a JS file from outside the entry-point. There is the case where an external import is a .d.ts file that then relatively imports `.d.ts` files of its own, but this is already covered by the `containingFile.endsWith('.js')` check.

JoostK replied:
I think you are correct that we can get rid of the check altogether. I put it in for the case where an import in an external package should not be rewritten to the `.js` file, but with the `endsWith('.js')` check this is alleviated.

There's one thing but I don't think it can actually happen: an external library import is resolved to a `.js` file, then one of its relative imports gets resolved to a `.d.ts` file. The latter step is weird, because the package did not get resolved to a `.d.ts` file in the first place so neither should its relative imports (as those must point into the same entry-point, which apparently does not have typings).

I will remove the check and verify that the tests still pass.

petebacondarwin replied:
Also this check would not be sufficient in the case where there was a `node_modules` within the entry-point folder :-)

I think it would be very weird for an external entry-point to resolve to a `.js` file and then also contain `.d.ts` files. In the `NewEntryPointFileWriter` I used the `entryPoint.package` path, to decide which files to copy. If we really needed to this check then that would be the better path to use - although there could still be weird corner cases between secondary entry-points.

petebacondarwin commented on packages/compiler-cli/ngcc/src/packages/ngcc_compiler_host.ts:
```diff
@@ -64,3 +64,43 @@ export class NgccCompilerHost implements ts.CompilerHost {
+      // JavaScript being present in the program. This logic recognizes this scenario and rewires
+      // the resolved .d.ts declaration file to its .js counterpart, if it exists.
+      if (resolvedModule !== undefined && resolvedModule.extension === ts.Extension.Dts &&
+          containingFile.endsWith('js') &&
```
```suggestion
          containingFile.endsWith('.js') &&
```

To be on the safe side :-)

JoostK replied:
Whoops, nice catch!

petebacondarwin reviewed (changes requested):
I agree that using `ModuleResolver` from the dependencies folder is not correct here. That class is optimised for dependency resolution rather than actually compiling a JS program.

Can you confirm whether you agree about the `entryPointPath` check not being needed and then tidy up the other nits.

Otherwise LGTM! NIce work.

petebacondarwin committed:
fix(ivy): use `CompilerHost.resolveModuleNames()` if available

Sometimes we need to override module resolution behaviour.
We do this by implementing the optional method `resolveModuleNames()`
on `CompilerHost`.

This commit ensures that we always try this method first before falling
back to the standard `ts.resolveModuleName`

petebacondarwin committed:
fixup! fix(ivy): ngcc - prefer JavaScript source files when resolving module imports

petebacondarwin commented:
Caretaker (merge-assistance): @JoostK  and @petebacondarwin have boths signed CLAs.

kara commented:
[presubmit](http://test/OCL:246185704:BASE:246185723:1556741718723:42548c3d)
